### ISSUE TITLE
Modify clocking strategy of EVC model

### ISSUE THREAD

**Issue by stefan-karg** [openETCS/modeling#535]
Some of the models in the EVC kernel need more than one cycle to generate valid outputs. In the modeling rules, the concept of "ready"-flags is introduced and used by some models.

At the moment, the ready flags are not used in the EVC integration and the environment model.

In order to get the EVC model running in a correct way, the clocking strategy of both the EVC and the environment model has to be modified.

To my knowledge, the ready flag is used in the following modules:

* MoRC
* InformationFilter (see #491)

**Comment by JakobGartner** [comment]
Hmm, 
ready flags.

For the MoRC I have no idea what they serve for, but for the Information Filter I have a very clear idea- they serve no purpose at all. 
Or to say it the other way round, most parts of the model don’t give a valid result at every cycle.

Discussing the information filter:

it can have two kinds of output:

a message
no message

we don’t care at all why it outputs *no* message at a given cycle.

It could be two reasons:

- the system has not received a message
- the input channel is processing a message (for example, it still did not yet receive all balises of a BG and can therefore not yet compose a BG message), or the filter needs some more cycles to receive and process all relevant information

Bottom line, the information filter will output a message with its “valid” flag as soon as it is available. 
We do NOT need an additional “ready” flag because our system is synchronised on the message level. 


Bottom line: We must make sure that the only “glue” between the components is the information that is flowing, which can be either based on the messages (until the track atlas and the M&L as well as the train position. These modules must each provide a clear state of whatever they calculate at every cycle. And if there is no information (yet) available at their outputs, we don’t care at all whether it is because they are still calculating or have not yet received the relevant info from the track.


I strongly oppose any attempt to introduce additional dependencies. I will jump out of the window if necessary. 


Jakob

**Comment by T12z** [comment]
... But with jumping out that window you could tear the flag down Super Mario-end-of-level-style ;-) 

And I agree with Jakob. The only reason to actually split a calculation across cycles would be specific optimisation which is not our goal.

---- Jakob Gärtner schrieb ----

>Hmm, 
>ready flags.
>
>For the MoRC I have no idea what they serve for, but for the Information Filter I have a very clear idea- they serve no purpose at all. 
>Or to say it the other way round, most parts of the model don’t give a valid result at every cycle.
>
>Discussing the information filter:
>
>it can have two kinds of output:
>
>a message
>no message
>
>we don’t care at all why it outputs *no* message at a given cycle.
>
>It could be two reasons:
>
>- the system has not received a message
>- the input channel is processing a message (for example, it still did not yet receive all balises of a BG and can therefore not yet compose a BG message), or the filter needs some more cycles to receive and process all relevant information
>
>Bottom line, the information filter will output a message with its “valid” flag as soon as it is available. 
>We do NOT need an additional “ready” flag because our system is synchronised on the message level. 
>
>
>Bottom line: We must make sure that the only “glue” between the components is the information that is flowing, which can be either based on the messages (until the track atlas and the M&L as well as the train position. These modules must each provide a clear state of whatever they calculate at every cycle. And if there is no information (yet) available at their outputs, we don’t care at all whether it is because they are still calculating or have not yet received the relevant info from the track.
>
>
>I strongly oppose any attempt to introduce additional dependencies. I will jump out of the window if necessary. 
>
>
>Jakob 
>
>
>

**Comment by BerndHekele** [comment]
@UweSteinkeFromSiemens 
Maybe you are interested in the discussion.

We have the ready flag in the MoRC and we need to handle it anyway. Can you please argue why  you don't like it? At least, it is recommended by our guide.

**Comment by UweSteinkeFromSiemens** [comment]
This topic actually arised because the ***InformationFilter*** requires the ability to feed internally stored datagrams into the regular datastream of messages from track. During this phase, there must not applied messages from track to the model concurrently. The ready signal can be used to provide additional execution cycles for the informationFilter without a conflict with a fresh message from track.
I don't contradict to any other solution that guarantees no more than one message at once for the subsequent model components after "ManageTrackSideInfomation_Integration".

***MoRC*** comprises internal feedback loops causing more than one cycle until the results are available at the ouputs in some situations. MoRC indicates this by deactivating it's ready output. While the ready output is unset, it's forbidden to apply new events to the MoRC inputs.  

In addition, **jumping out of the window** is strongly prohibited as long as the ready signal is unset.

**Comment by JakobGartner** [comment]
After analysis of the modules in question:

Our model has an architecture where
- the top level (=root node) is being executed at a fixed cycle time (we imagine 50ms at the moment) which makes sure that we can react to events at a rate of 20Hz.
- The communication with the outside world must be compatible with the message- based concept of ETCS. 

We are now talking about two modules where we have use of what is called a Ready Flag.

After analysis, here a brief summary of the findings:

- The **message filter** may have the need to delay some messages while making sure that the overall sequencing of messages stays consistent, while obeying the overall restrictions on processing time.

I consider this to be a local problem which must be encapsulated in the filter, or maybe with the filter plus some local components of the  EVC Track to Train frontend. 

The rest of the system *does not want to know* what happens before a message with its packet leaves the output of the filter. 
The only thing that needs to be known:
- There is a valid message at this cycle (which at the moment is indicated by its 'valid' flag
- The variables and packets that come along with the message

The Ork (sorry, **MoRC**) is a classical state machine: it reacts to a sequence of events and may or may not reach an internal state that then will make it emit some information at an unknown time.

The coupling with other parts of the system is again by a set of inputs, each of which will have a defined value at each cycle. Some of which may be indeed abstractions of messages (indicating by a 'valid' flag that they carry information in a given cycle.

On the output side, the MoRC emits state variables and/ or messages (with their 'valid' flags)

The *world outside of the Ork* does not care about its internal problems:
There is a data flow that the MoRC must digest at every cycle
There is a protocol that it has with several counterparts which is defined by the SRS (on the various layers of communication)
Whatever "ready flag" it may want to have internally, it must never assume that the outside world is interested in it.
Hence, it must handle any necessary buffering on in- and output sides internally. 

In realiy I don't believe that this can be a real problem. Every state machine works like this. It reacts on a sequence of the input state space and produces a sequence of the output space.

Thanks to SCADE, it can never see deadlocks, circular dependencies, ambiguities.

As an example of state machines that have the same problem (they will only need to emit a signal to the outside world at a time which is unknown and have to cope with a defined, but in detail unknown sequence of their input space) we have all procedures and also the mode and level management which perfectly live without a ready flag. 
ETCS language provides the necessary elements for flow control in form of the Acks and timers.

**Comment by UweSteinkeFromSiemens** [comment]
The "ready" flag concept has been introduced in general with https://github.com/openETCS/modeling/blob/master/ModelingRules/SCADE_Modelling_Guide.pdf at beginning 2014. It suits to the message based concept of ETCS and - in addition - minimizes the systems response time and response time deviation to mainly 1 major cycle period (≈ 300 ms for the Alstom API).  

The ready flag is foreseen for the environment of a function, not for the function itself. Not to respect the ready flag by not generationg extra clocks while the ready flag is unset, will simply lead to extra clock periods until some of the results are provided and so prolong the response time. 

The second benefit of the ready flag concept is that a function can rely on not to receive a new event at its inputs as long as the previous one has been consumed completely. This avoids concurrent and competitive signal racing situations systematically. 

MoRC does not require the ready flag for its own. Without taking the ready flag into account, subsequent functions and the OBU models environment only have to consider the extra periods it may take until the results are available.

Without the ready flag, the InformationFilter has to buffer input events it receives during the phase of feeding its internal message store into the data stream. 

In summary, deviating from the ready flag concept increases the complexity of single functions as well as of the whole system, due to precautions to be implemented additionally inside dedicated functions and prolonged and increased varied response times.
